parrot.rs

```diff
@@ -0,0 +1,73 @@
+use core::sync::atomic::{AtomicUsize, Ordering};
+use kernel::{
+    bindings::msleep,
+    chrdev::Registration,
+    file::{File, Operations, ToUse},
+    io_buffer::IoBufferWriter,
+    prelude::*,
+};
+
+mod frames;
+use frames::*;
+
+module! {
+    type: ParrotSafe,
+    name: b"party_parrot",
+    author: b"John Baublitz (john.m.baublitz@gmail.com)",
+    description: b"Get the party started",
+    license: b"GPL",
+}
+
+const FRAMES: [&str; 10] = [
+    FRAME0, FRAME1, FRAME2, FRAME3, FRAME4, FRAME5, FRAME6, FRAME7, FRAME8, FRAME9,
+];
+
+struct ParrotSafe(Pin<Box<Registration<1>>>);
+
+impl kernel::Module for ParrotSafe {
+    fn init(name: &'static CStr, module: &'static ThisModule) -> Result<Self> {
+        let mut registration = Registration::new_pinned(name, 0, module)?;
+        registration.as_mut().register::<ParrotOps>()?;
+        Ok(ParrotSafe(registration))
+    }
+}
+
+struct ParrotOps(usize);
+
+impl<'a> Operations for ParrotOps {
+    const TO_USE: ToUse = ToUse {
+        read: true,
+        read_iter: false,
+        write: false,
+        write_iter: false,
+        compat_ioctl: false,
+        fsync: false,
+        ioctl: false,
+        mmap: false,
+        poll: false,
+        seek: false,
+    };
+
+    type Data = Box<AtomicUsize>;
+
+    fn open(_: &Self::OpenData, _: &File) -> Result<Self::Data> {
+        Ok(Box::try_new(AtomicUsize::new(0))?)
+    }
+
+    fn read(data: &AtomicUsize, _: &File, buf: &mut impl IoBufferWriter, _: u64) -> Result<usize> {
+        let frame = FRAMES
```

Comment:
Why not use `fetch_add` (which doesn't fail) and do the modulo operation locally? Something like:

```rust
let frame = FRAMES.get(data.fetch_add(1, Ordering::Relaxed) % FRAMES.len());
```

Reply:
From https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.fetch_add:

> This operation wraps around on overflow.

Reply:
Oh perfect, I'll make that change.